From 69addbc16d05daf02bf5f9a4ddd3262158517b1c Mon Sep 17 00:00:00 2001
From: Meriipu <Meriipu@users.noreply.github.com>
Date: Fri, 17 Dec 2021 15:55:13 +0100
Subject: [PATCH 03/10] revert the removal of the dropmarker et al.

as of 78 this alone is not enough to have the dropmarker be functional
some additional patches to fix autoOpen on focus/select as well as
avoiding one of the providers is necessary. This commit alone only adds
the dropmarker back cosmetically.
---
 browser/base/content/browser.css              |  1 +
 .../base/content/navigator-toolbox.inc.xhtml  |  3 +++
 browser/components/urlbar/UrlbarInput.jsm     | 19 +++++++++++++++
 browser/components/urlbar/UrlbarView.jsm      |  3 +++
 .../themes/shared/urlbar-searchbar.inc.css    | 24 +++++++++++++++++++
 5 files changed, 50 insertions(+)

diff --git a/browser/base/content/browser.css b/browser/base/content/browser.css
index 2d74162b1543..8d8abf5e8b59 100644
--- a/browser/base/content/browser.css
+++ b/browser/base/content/browser.css
@@ -730,6 +730,7 @@ toolbar:not(#TabsToolbar) > #personal-bookmarks {
 #urlbar-input-container[pageproxystate=invalid] > #page-action-buttons > .urlbar-page-action,
 #identity-box.chromeUI ~ #page-action-buttons > .urlbar-page-action:not(#star-button-box),
 #urlbar[usertyping] > #urlbar-input-container > #page-action-buttons > #urlbar-zoom-button,
+#urlbar[usertyping] > #urlbar-input-container > .urlbar-history-dropmarker,
 #urlbar:not([usertyping]) > #urlbar-input-container > #urlbar-go-button,
 #urlbar:not([focused]) > #urlbar-input-container > #urlbar-go-button {
   display: none;
diff --git a/browser/base/content/navigator-toolbox.inc.xhtml b/browser/base/content/navigator-toolbox.inc.xhtml
index 02636a6b46b5..d683e3e6504d 100644
--- a/browser/base/content/navigator-toolbox.inc.xhtml
+++ b/browser/base/content/navigator-toolbox.inc.xhtml
@@ -328,6 +328,9 @@
                    class="urlbar-icon"
                    onclick="gURLBar.handleCommand(event);"
                    data-l10n-id="urlbar-go-button"/>
+            <image class="urlbar-history-dropmarker urlbar-icon chromeclass-toolbar-additional"
+                   hidden="false"
+                   tooltiptext="Show history"/>
             <hbox id="page-action-buttons" context="pageActionContextMenu">
               <toolbartabstop/>
               <hbox id="contextual-feature-recommendation" role="button" hidden="true">
diff --git a/browser/components/urlbar/UrlbarInput.jsm b/browser/components/urlbar/UrlbarInput.jsm
index 420ff49a09e3..33130d51054a 100644
--- a/browser/components/urlbar/UrlbarInput.jsm
+++ b/browser/components/urlbar/UrlbarInput.jsm
@@ -182,6 +182,8 @@ class UrlbarInput {
     }
 
     this.inputField = this.querySelector("#urlbar-input");
+    this.dropmarker = this.querySelector(".urlbar-history-dropmarker");
+    this.dropmarker.hidden = false;
     this._inputContainer = this.querySelector("#urlbar-input-container");
     this._identityBox = this.querySelector("#identity-box");
     this._searchModeIndicator = this.querySelector(
@@ -240,6 +242,7 @@ class UrlbarInput {
       this.addEventListener(name, this);
     }
 
+    this.dropmarker.addEventListener("mousedown", this);
     this.window.addEventListener("mousedown", this);
     if (AppConstants.platform == "win") {
       this.window.addEventListener("draggableregionleftmousedown", this);
@@ -2904,6 +2907,22 @@ class UrlbarInput {
           });
         }
         break;
+      case this.dropmarker:
+        if (event.button != 0) {
+          break;
+        }
+
+        if (this.view.isOpen) {
+          this.view.close();
+        } else {
+          this.focus();
+          this.startQuery({
+            allowAutofill: false,
+            event,
+          });
+          this._maybeSelectAll();
+        }
+        break;
       case this.window:
         if (this._mousedownOnUrlbarDescendant) {
           this._mousedownOnUrlbarDescendant = false;
diff --git a/browser/components/urlbar/UrlbarView.jsm b/browser/components/urlbar/UrlbarView.jsm
index 428c9e370a98..2c6f4a6eca27 100644
--- a/browser/components/urlbar/UrlbarView.jsm
+++ b/browser/components/urlbar/UrlbarView.jsm
@@ -462,6 +462,8 @@ class UrlbarView {
     this._openPanelInstance = null;
     this._previousTabToSearchEngine = null;
 
+    this.input.dropmarker.removeAttribute("open");
+
     this.input.removeAttribute("open");
     this.input.endLayoutExtend();
 
@@ -871,6 +873,7 @@ class UrlbarView {
     this._enableOrDisableRowWrap();
 
     this.input.inputField.setAttribute("aria-expanded", "true");
+    this.input.dropmarker.setAttribute("open", "true");
 
     this.input.toggleAttribute("suppress-focus-border", true);
     this.input.setAttribute("open", "true");
diff --git a/browser/themes/shared/urlbar-searchbar.inc.css b/browser/themes/shared/urlbar-searchbar.inc.css
index 82675dae2041..eba08d805e0d 100644
--- a/browser/themes/shared/urlbar-searchbar.inc.css
+++ b/browser/themes/shared/urlbar-searchbar.inc.css
@@ -453,6 +453,19 @@
   transform: scaleX(-1);
 }
 
+.urlbar-history-dropmarker {
+  list-style-image: url(chrome://global/skin/icons/arrow-dropdown-16.svg);
+  transition: opacity 0.15s ease;
+}
+
+#urlbar[switchingtabs] > #urlbar-input-container > .urlbar-history-dropmarker {
+  transition: none;
+}
+
+#nav-bar:not([customizing="true"]) > #nav-bar-customization-target > #urlbar-container:not(:hover) > #urlbar:not([focused]) > #urlbar-input-container > .urlbar-history-dropmarker {
+  opacity: 0;
+}
+
 #pageActionButton,
 .share-more-button {
   list-style-image: url("chrome://global/skin/icons/more.svg");
@@ -572,6 +585,17 @@
   mask-position-x: 0;
 }
 
+/* Translate the dropmarker to give illusion of increasing width of the recommendation  */
+#urlbar[cfr-recommendation-state] .urlbar-history-dropmarker {
+  transition: transform ease-in-out var(--cfr-animation-duration);
+}
+#urlbar[cfr-recommendation-state="expanded"] .urlbar-history-dropmarker {
+  transform: translateX(calc(var(--cfr-label-width) * -1));
+}
+#urlbar[cfr-recommendation-state="expanded"] .urlbar-history-dropmarker:-moz-locale-dir(rtl) {
+  transform: translateX(calc(var(--cfr-label-width)));
+}
+
 /* Shift the url-bar text fading to stop the recommendation overlapping */
 #urlbar[cfr-recommendation-state] #urlbar-input {
   /* A mask-image is usually only defined for the url bar when text overflows.
-- 
2.34.1

